pkg/fakeclient/client.go

```diff
@@ -0,0 +1,417 @@
+	}
+
+	for event := range t.ResultChan() {
+		t.dispatch(&event)
+	}
+
+	return nil
+}
+
+func (t *FakeObjectTracker) dispatch(event *watch.Event) {
+	for _, w := range t.watchers {
+		go w.dispatch(event)
+	}
+}
+
+// Stop terminates tracking of an FakeObjectTracker
+func (t *FakeObjectTracker) Stop() {
+	if t.FakeWatcher == nil {
+		panic(errors.New("Tracker has no watch support"))
+	}
+
+	t.trackerMutex.Lock()
+	defer t.trackerMutex.Unlock()
+
+	t.FakeWatcher.Stop()
+	for _, w := range t.watchers {
+		w.Stop()
+	}
+	t.watchers = []*watcher{}
+}
+
+type watcher struct {
+	*watch.FakeWatcher
+	action      k8stesting.WatchAction
+	updateMutex sync.Mutex
+}
+
+func (w *watcher) Stop() {
+	w.updateMutex.Lock()
+	defer w.updateMutex.Unlock()
+
+	w.FakeWatcher.Stop()
+}
+
+func (w *watcher) handles(event *watch.Event) bool {
+	if w.IsStopped() {
+		return false
+	}
+
+	t, err := meta.TypeAccessor(event.Object)
+	if err != nil {
+		return false
+	}
+
+	gvr, _ := meta.UnsafeGuessKindToResource(schema.FromAPIVersionAndKind(t.GetAPIVersion(), t.GetKind()))
+	if !(&k8stesting.SimpleWatchReactor{Resource: gvr.Resource}).Handles(w.action) {
+		return false
+	}
+
+	o, err := meta.Accessor(event.Object)
+	if err != nil {
+		return false
+	}
+
+	info := w.action.GetWatchRestrictions()
+	rv, fs, ls := info.ResourceVersion, info.Fields, info.Labels
+	if rv != "" && o.GetResourceVersion() != rv {
+		return false
+	}
+
+	if fs != nil && !fs.Matches(fields.Set{
+		"metadata.name":      o.GetName(),
+		"metadata.namespace": o.GetNamespace(),
+	}) {
+		return false
+	}
+
+	if ls != nil && !ls.Matches(labels.Set(o.GetLabels())) {
+		return false
+	}
+
+	return true
+}
+
+func (w *watcher) dispatch(event *watch.Event) {
+	w.updateMutex.Lock()
+	defer w.updateMutex.Unlock()
+
+	if !w.handles(event) {
+		return
+	}
+	w.Action(event.Type, event.Object)
+}
+
+func (w *watcher) dispatchInitialObjects(action k8stesting.WatchAction, t k8stesting.ObjectTracker) error {
+	listObj, err := t.List(action.GetResource(), action.GetResource().GroupVersion().WithKind(action.GetResource().Resource), action.GetNamespace())
+	if err != nil {
+		return err
+	}
+
+	itemsPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+
+	items := itemsPtr.([]runtime.Object)
+	for _, o := range items {
+		w.dispatch(&watch.Event{
+			Type:   watch.Added,
+			Object: o,
+		})
+	}
+
+	return nil
+}
+
+// fakingOptions are options that can be set while trying to fake object tracker returns
+type fakingOptions struct {
+	fakingEnabled bool
+	errorMessage  string
+	delay         time.Duration
+}
+
+// SetDelay sets delay while invoking any interface exposed by standard ObjectTrackers
+func (o *fakingOptions) SetDelay(delay time.Duration) error {
+	o.fakingEnabled = true
+	o.delay = delay
+	return nil
+}
+
+// SetError sets up the errorMessage to be returned on further function calls
+func (o *fakingOptions) SetError(message string) error {
```

Comment:
For the time being this is OK. In the future, we can think of adding more flexibility to return the error only if certain conditions are met.

Reply:
Sure, we shall add/modify methods as and when required.